### ISSUE TITLE
Taking a shared reference of an array suppresses the unconditional_panic lint

### ISSUE THREAD

**Issue by JanBeh** [rust-lang/rust#98444]
I tried this code:

```rust
fn main() {
    let xs: [i32; 5] = [1, 2, 3, 4, 5];
    let _ = &xs; // this line suppresses the `unconditional_panic` lint
    let _ = xs[7];
}
```

I expected to see this happen:

```text
error: this operation will panic at runtime
 --> src/main.rs:4:13
  |
4 |     let _ = xs[7];
  |             ^^^^^ index out of bounds: the length is 5 but the index is 7
  |
  = note: `#[deny(unconditional_panic)]` on by default
```

Instead, this happened:

```text
thread 'main' panicked at 'index out of bounds: the len is 5 but the index is 7', src/main.rs:4:13
```

Apparently, taking a shared reference of `xs` will suppress the `unconditional_panic` lint. This seems like a false negative. While false negatives can't be avoided entirely, this case seems to be avoidable as taking a shared reference shouldn't change the type of `xs` (which is `[i32; 5]`) or modify any data in this example.

Note that

```rust
fn main() {
    let xs: [i32; 5] = [1, 2, 3, 4, 5];
    //let _ = &xs;
    let _ = xs[7];
}
```

will trigger the lint properly.

### Meta
`rustc --version --verbose`:
```
rustc 1.63.0-nightly (43347397f 2022-06-23)
binary: rustc
commit-hash: 43347397f7c5ca9a670a3bb3890c7187e24a52ab
commit-date: 2022-06-23
host: x86_64-unknown-freebsd
release: 1.63.0-nightly
LLVM version: 14.0.5
```

**Comment by JanBeh** [comment]
See also [this thread on URLO](https://users.rust-lang.org/t/out-of-bound-compile-error-with-len-function-call/77342).

**Comment by poliorcetics** [comment]
Note this is a problem because [the book has an example](https://doc.rust-lang.org/rust-by-example/primitives/array.html) which is supposed to fail to compile but this bug makes it fail at runtime instead

**Comment by kmdreko** [comment]
In case no one has dug into it yet, the error is suppressed in `1.61` but not in `1.60` and before:

```
$ cargo +1.61 build
   Compiling mycrate v0.1.0 (/rust-tests)
    Finished dev [unoptimized + debuginfo] target(s) in 0.74s
```
```
$ cargo +1.60 build
   Compiling mycrate v0.1.0 (/rust-tests)
error: this operation will panic at runtime
   --> src/main.rs:129:13
    |
129 |     let _ = xs[7];
    |             ^^^^^ index out of bounds: the length is 5 but the index is 7
    |
    = note: `#[deny(unconditional_panic)]` on by default

error: could not compile `mycrate` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2398444.20Taking.20a.20shared.20reference.20of.20an.20array.20suppresses.20the.E2.80.A6)).

@rustbot label -I-prioritize +P-medium

**Comment by TheWastl** [comment]
This is caused by #94934. It only ever worked because the lint ran after optimizations.

I made a fix [here](https://github.com/TheWastl/rust/commit/090872eb3d6ca04ed5c5ed78af224c5648c85da0). It specializes the length of arrays as constant, even if the array isn't. I'm going to open a PR when I have time, probably tomorrow.

In the long term, I'd like to implement some more extensive tracking of "what parts of a value are constant". For example, the compiler currently assumes that anything that had its reference taken can change, but actually, this can only happen inside `UnsafeCell` (as I understand it).

**Comment by steffahn** [comment]
Labels from the closed #101470, @rustbot label A-diagnostics, A-lint, A-mir-opt, T-compiler.

**Comment by cjgillot** [comment]
Fixed on stable
@rustbot label E-needs-test

**Comment by albertlarsan68** [comment]
I claim the add test part

**Comment by albertlarsan68** [comment]
This lint only appears on full builds (not check), but this will be fixed once #108730 lands.

**Comment by apiraino** [comment]
Reopening because I feel this is not yet solved, see [Zulip notes](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23114840.20Compile.20time.20bounds.20checking.20doesn't.20work.20anymore.20f.E2.80.A6) on #114840.

As per the [prev. comment](https://github.com/rust-lang/rust/issues/98444#issuecomment-1456155660), #108730 should nail down this issue?

**Comment by veera-sivarajan** [comment]
MRE:
``` rust
fn main() {
    let slice = &[0, 1];
    let _ = slice[2];
}
```

**Comment by veera-sivarajan** [comment]
Another variant of this same issue: #109260
